### PR TITLE
[WIP] *: dump snapshot to file

### PR DESCRIPTION
**Pull request by tiancaiamao**
I think I should not change kvproto and Snapshot format, at least not for now.
I think I make things complex than it should be.

### CODE REVIEW

File: src/raftstore/store/worker/snap.rs [hunk 1]
```diff
@@ -70,16 +88,79 @@ impl Runner {
...
+    }
+
+    fn save_snapshot(&self, snapshot: &Snapshot, dir: &str) -> Result<(), Error> {
+        let header = snapshot.get_metadata();
+        let file_name = format!("{}{}_{}_{}",
+                                dir,
+                                header.get_region().get_id(),
+                                header.get_term(),
+                                header.get_index());
+        let f = box fs::File::create(format!("{}xxx.tmp", dir)).unwrap();
```
Comment (siddontang) [inline src/raftstore/store/worker/snap.rs]: why "xxx"?

Comment (siddontang) [inline src/raftstore/store/worker/snap.rs]: The name "xxx" is bad, remove.

Comment (tiancaiamao) [inline src/raftstore/store/worker/snap.rs]: Agree

Comment (siddontang) [inline src/raftstore/store/worker/snap.rs]: should check whether file exists already.

File: src/raftstore/store/peer_storage.rs [hunk 2]
```diff
@@ -523,6 +523,7 @@ pub fn do_snapshot(snap: &RocksDbSnapshot,
     // Set snapshot metadata.
     snapshot.mut_metadata().set_index(applied_idx);
     snapshot.mut_metadata().set_term(term);
+    snapshot.mut_metadata().set_region(region.clone());
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: remove this, internal raft doesn't know any region.

File: src/raftstore/store/worker/snap.rs [hunk 3]
```diff
@@ -70,16 +88,81 @@ impl Runner {
...
+        let res = self.generate_snap(&task);
+        if let Err(e) = res {
             error!("failed to generate snap: {:?}!!!", e);
             task.storage.wl().snap_state = SnapState::Failed;
+            return;
         }
+
+        let (snap, region_id) = res.unwrap();
+        if let Err(e) = self.save_snapshot(&snap, region_id, "/tmp/") {
+            error!("save snapshot file failed: {:?}!!!", e);
```
Comment (siddontang) [inline src/raftstore/store/worker/snap.rs]: set snap state failed here.

File: src/raftstore/store/worker/snap.rs [hunk 4]
```diff
@@ -70,16 +88,81 @@ impl Runner {
...
+        }
+        task.storage.wl().snap_state = SnapState::Snap(snap);
+    }
+}
+
+struct CRCWriter {
+    digest: crc32::Digest,
+    writer: Box<Write>,
```
Comment (siddontang) [inline src/raftstore/store/worker/snap.rs]: no need box here.

File: src/raftstore/store/worker/snap.rs [hunk 5]
```diff
@@ -70,16 +88,83 @@ impl Runner {
...
-        if let Err(e) = self.generate_snap(&task) {
+        let res = self.generate_snap(&task);
+        if let Err(e) = res {
             error!("failed to generate snap: {:?}!!!", e);
             task.storage.wl().snap_state = SnapState::Failed;
+            return;
+        }
+
+        let (snap, region_id) = res.unwrap();
+        if let Err(e) = self.save_snapshot(&snap, region_id, "/tmp/") {
```
Comment (siddontang) [inline src/raftstore/store/worker/snap.rs]: please use a configuration for the snapshot path.

**Comment by tiancaiamao** [comment]
will be include in https://github.com/pingcap/tikv/pull/648

**Pull request closed without merging**